content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,526 @@
+
+#### `select()`
+
+The dplyr code:
+
+```{r select, eval = FALSE}
+select(mtcars_tbl, hp)
+```
+
+...can be replaced by:
+
+```{r select-alt, eval = FALSE}
+mtcars_tbl["hp"]
+```
+
+...with benchmarks:
+
+```{r bm-select}
+bench::mark(
+  dplyr      = select(mtcars_tbl, hp),
+  `[.tbl_df` = mtcars_tbl["hp"]
+) %>%
+  select(expression, median)
+```
+
+Of course, the nice part about `select()`, and something we make use of in tidymodels quite a bit, is tidyselect. I've often found that we lean heavily on selecting via external vectors, i.e. character vectors, i.e. things that can be inputted to `[.tbl_df` directly. That is:
+
+```{r bm-select-2}
+cols <- c("hp", "wt")
+
+bench::mark(
+  dplyr = select(mtcars_tbl, all_of(cols)),
+  `[.tbl_df` = mtcars_tbl[cols]
+) %>% 
+  select(expression, median)
+```
+
+Note that `[.tbl_df` always sets `drop = FALSE`.
+
+`[.tbl_df` can also be used as an alternative interface to `select()` with a `.before` or `.after` argument as well as `relocate()`. For instance, to place that column `year` we made in the last section as the second column, we could write:
+
+```{r left-cols, include = FALSE}
+left_cols <- c("make_model", "year")
+```
+
+```{r left-cols-select, eval = FALSE}
+left_cols <- c("make_model", "year")
+
+mtcars_tbl[
+  c(left_cols, 
+    setdiff(colnames(mtcars_tbl), left_cols)
+  )
+]
+```
+
+No, thanks, but it is a good bit faster than tidyselect-based alternatives:
+
+```{r bm-relocate}
+bench::mark(
+  mutate = mutate(mtcars_tbl, year = 1974L, .after = make_model),
+  relocate = relocate(mtcars_tbl, year, .after = make_model),
+  `[.tbl_df` = 
+      mtcars_tbl[
+        c(left_cols, 
+          colnames(mtcars_tbl[!colnames(mtcars_tbl) %in% left_cols])
+        )
+      ],
+  check = FALSE
+) %>% 
+  select(expression, median)
+```
+
+```{r null-year, include = FALSE}
+mtcars_tbl$year <- NULL
+```
+
+Some examples of changes like this made to tidymodels packages: [tidymodels/parsnip#935](https://github.com/tidymodels/parsnip/pull/935), [tidymodels/parsnip#933](https://github.com/tidymodels/parsnip/pull/933), [tidymodels/parsnip#921](https://github.com/tidymodels/parsnip/pull/921), and [tidymodels/tune#635](https://github.com/tidymodels/tune/pull/635).
+
+#### `pull()`
+
+The dplyr code:
+
+```{r pull, eval = FALSE}
+pull(mtcars_tbl, hp)
+```
+
+...can be replaced by:
+
+```{r pull-alt, eval = FALSE}
+mtcars_tbl$hp
+```
+
+...or:
+
+```{r pull-alt-2, eval = FALSE}
+mtcars_tbl[["hp"]]
+```
+
+Note that `$` will raise a warning if the subsetted column doesn't exist, while `[[` will silently return `NULL`.
+
+With benchmarks:
+
+```{r bm-pull}
+bench::mark(
+  dplyr      = pull(mtcars_tbl, hp),
+  `$.tbl_df` = mtcars_tbl$hp,
+  `[[.tbl_df` = mtcars_tbl[["hp"]]
```

Comment:
Alignment of `=` is off. In general i dont think there is too much benefit to trying to add extra spaces to keep the RHSs aligned